src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioShutdownCancellationService.cs

```diff
@@ -0,0 +1,27 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Composition;
+using System.Threading;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.Shell;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation
+{
+    [ExportWorkspaceService(typeof(IShutdownCancellationService), ServiceLayer.Host), Shared]
+    internal sealed class VisualStudioShutdownCancellationService : IShutdownCancellationService
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public VisualStudioShutdownCancellationService()
+        {
+        }
+
+        public CancellationToken ShutdownToken => VsShellUtilities.ShutdownToken;
```

Comment:
interesting.  what's the diff between this and the .DIsposalToken you can get from IThreadingContext?  if they're the same, can we unify? if they're different we should doc when you should use one versus the other.

Reply:
I don't know where IThreadingContext.DisposalToken comes from. `VsShellUtilities.ShutdownToken` is signaled when VS starts shutting down.

Reply:
@AArnott Is `IThreadingContext.DisposalToken` different from `VsShellUtilities.ShutdownToken`?

Reply:
@sharwell as well.  I imagine we'll want to unify on one strategy here whatever we do.  Unless they truly represent separate concepts and we need to care about those differences (in which case we need to doc well) :)

Reply:
I don't know where `IThreadingContext.DisposalToken` comes from. My guess though is it is associated with a VS package. If that's true, it's signaled later than the VsShellUtilities one.

Reply:
@AArnott Turns out this is actually Roslyn type wrapping JTF I did not realize that. I thought it's coming directly from VS somehow.

Reply:
Seems like the token is signaled when the package is being closed:

```
>	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Shared.Utilities.ThreadingContext.Dispose() Line 75	C#
 	Microsoft.VisualStudio.Composition.dll!Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.Dispose() Line 1106	C#
 	Microsoft.VisualStudio.Composition.dll!Microsoft.VisualStudio.Composition.ExportProvider.Dispose(bool disposing) Line 416	C#
 	Microsoft.VisualStudio.ComponentModelHost.Implementation.dll!Microsoft.VisualStudio.ComponentModelHost.ComponentModel.Dispose(bool disposing) Line 308	C#
 	Microsoft.VisualStudio.ComponentModelHost.Implementation.dll!Microsoft.VisualStudio.ComponentModelHost.ComponentModel.Dispose() Line 281	C#
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.AsyncPackage.RemoveService(System.Type serviceType, bool ignored) Line 602	C#
 	Microsoft.VisualStudio.ComponentModelHost.Implementation.dll!Microsoft.VisualStudio.ComponentModelHost.HostPackage.Dispose(bool disposing) Line 52	C#
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.Package.Microsoft.VisualStudio.Shell.Interop.IVsPackage.Close() Line 1688	C#
```